
### Sentiment Analysis via the ML-based approach



### Loading and Prep



In [ ]:
# import necessary libraries

import pandas as pd
import numpy as np
import re
import pandas as pd
import nltk
import matplotlib.pyplot as plt


In [ ]:
# Load train dataset

from google.colab import files
uploaded = files.upload()

Saving sentiment_train.csv to sentiment_train (1).csv


In [ ]:
df_train = pd.read_csv("sentiment_train.csv")

print(df_train.info())
print(df_train.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2400 entries, 0 to 2399
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   Sentence  2400 non-null   object
 1   Polarity  2400 non-null   int64 
dtypes: int64(1), object(1)
memory usage: 37.6+ KB
None
                                            Sentence  Polarity
0                           Wow... Loved this place.         1
1                                 Crust is not good.         0
2          Not tasty and the texture was just nasty.         0
3  Stopped by during the late May bank holiday of...         1
4  The selection on the menu was great and so wer...         1


In [ ]:
# Load test dataset

from google.colab import files
uploaded = files.upload()

Saving sentiment_test.csv to sentiment_test (1).csv


In [ ]:
df_test = pd.read_csv("sentiment_test.csv")

print(df_test.info())
print(df_test.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 600 entries, 0 to 599
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   Sentence  600 non-null    object
 1   Polarity  600 non-null    int64 
dtypes: int64(1), object(1)
memory usage: 9.5+ KB
None
                                            Sentence  Polarity
0  A good commentary of today's love and undoubte...         1
1  For people who are first timers in film making...         1
2  It was very popular when I was in the cinema, ...         1
3  It's a feel-good film and that's how I felt wh...         1
4  It has northern humour and positive about the ...         1


####Polarity which is the target variable is categorized as either 0 or 1 for each sentence ( 0 as negative, 1 as positive)

In [ ]:

# check missing values in train data set
df_train.isnull().sum()

Sentence    0
Polarity    0
dtype: int64

In [ ]:
# check missing values in test data set
df_test.isnull().sum()

Sentence    0
Polarity    0
dtype: int64

Text Preprocessing on Training Data

In [ ]:
# Get everything from nltk library

nltk.download('all')

[nltk_data] Downloading collection 'all'
[nltk_data]    | 
[nltk_data]    | Downloading package abc to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/abc.zip.
[nltk_data]    | Downloading package alpino to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/alpino.zip.
[nltk_data]    | Downloading package averaged_perceptron_tagger to
[nltk_data]    |     /root/nltk_data...
[nltk_data]    |   Unzipping taggers/averaged_perceptron_tagger.zip.
[nltk_data]    | Downloading package averaged_perceptron_tagger_ru to
[nltk_data]    |     /root/nltk_data...
[nltk_data]    |   Unzipping
[nltk_data]    |       taggers/averaged_perceptron_tagger_ru.zip.
[nltk_data]    | Downloading package basque_grammars to
[nltk_data]    |     /root/nltk_data...
[nltk_data]    |   Unzipping grammars/basque_grammars.zip.
[nltk_data]    | Downloading package biocreative_ppi to
[nltk_data]    |     /root/nltk_data...
[nltk_data]    |   Unzipping corpora/biocreative_ppi.zip.
[nltk_data]    | Downloadin

True

In [ ]:
text = list(df_train['Sentence'])

### Text Pre-processing on Train Data


In [ ]:
# The first pre-processing step we’ll do is transform all sentences into lower case and create a new column sentence_lower.

df_train["sentence_lower"] = df_train["Sentence"].str.lower()
df_train.head()

Sentence  Polarity  \
0                           Wow... Loved this place.         1   
1                                 Crust is not good.         0   
2          Not tasty and the texture was just nasty.         0   
3  Stopped by during the late May bank holiday of...         1   
4  The selection on the menu was great and so wer...         1   

                                      sentence_lower  
0                           wow... loved this place.  
1                                 crust is not good.  
2          not tasty and the texture was just nasty.  
3  stopped by during the late may bank holiday of...  
4  the selection on the menu was great and so wer...

In [ ]:
# Remove punctuations

import string

PUNCT_TO_REMOVE = string.punctuation
def remove_punctuation(text):
    #custom function to remove the punctuation
    return text.translate(str.maketrans('', '', PUNCT_TO_REMOVE))

# Passing the function to 'Sentence' and storing into new column 'sentence_punc'

df_train["sentence_punc"] = df_train["Sentence"].apply(lambda text: remove_punctuation(text))
df_train.head()

Sentence  Polarity  \
0                           Wow... Loved this place.         1   
1                                 Crust is not good.         0   
2          Not tasty and the texture was just nasty.         0   
3  Stopped by during the late May bank holiday of...         1   
4  The selection on the menu was great and so wer...         1   

                                      sentence_lower  \
0                           wow... loved this place.   
1                                 crust is not good.   
2          not tasty and the texture was just nasty.   
3  stopped by during the late may bank holiday of...   
4  the selection on the menu was great and so wer...   

                                       sentence_punc  
0                               Wow Loved this place  
1                                  Crust is not good  
2           Not tasty and the texture was just nasty  
3  Stopped by during the late May bank holiday of...  
4  The selection on the menu was great and so wer...

In [ ]:
# remove stop words

from nltk.corpus import stopwords
", ".join(stopwords.words('english'))

"i, me, my, myself, we, our, ours, ourselves, you, you're, you've, you'll, you'd, your, yours, yourself, yourselves, he, him, his, himself, she, she's, her, hers, herself, it, it's, its, itself, they, them, their, theirs, themselves, what, which, who, whom, this, that, that'll, these, those, am, is, are, was, were, be, been, being, have, has, had, having, do, does, did, doing, a, an, the, and, but, if, or, because, as, until, while, of, at, by, for, with, about, against, between, into, through, during, before, after, above, below, to, from, up, down, in, out, on, off, over, under, again, further, then, once, here, there, when, where, why, how, all, any, both, each, few, more, most, other, some, such, no, nor, not, only, own, same, so, than, too, very, s, t, can, will, just, don, don't, should, should've, now, d, ll, m, o, re, ve, y, ain, aren, aren't, couldn, couldn't, didn, didn't, doesn, doesn't, hadn, hadn't, hasn, hasn't, haven, haven't, isn, isn't, ma, mightn, mightn't, mustn, mus

In [ ]:
STOPWORDS = set(stopwords.words('english'))
def remove_stopwords(text):
    """custom function to remove the stopwords"""
    return " ".join([word for word in str(text).split() if word not in STOPWORDS])

# Passing the function to 'sentence_punc' and storing into new column 'sentence_stopwords'
df_train["sentence_stopwords"] = df_train["sentence_punc"].apply(lambda text: remove_stopwords(text))
df_train.head()

Sentence  Polarity  \
0                           Wow... Loved this place.         1   
1                                 Crust is not good.         0   
2          Not tasty and the texture was just nasty.         0   
3  Stopped by during the late May bank holiday of...         1   
4  The selection on the menu was great and so wer...         1   

                                      sentence_lower  \
0                           wow... loved this place.   
1                                 crust is not good.   
2          not tasty and the texture was just nasty.   
3  stopped by during the late may bank holiday of...   
4  the selection on the menu was great and so wer...   

                                       sentence_punc  \
0                               Wow Loved this place   
1                                  Crust is not good   
2           Not tasty and the texture was just nasty   
3  Stopped by during the late May bank holiday of...   
4  The selection on the menu was great and so wer...   

                                  sentence_stopwords  
0                                    Wow Loved place  
1                                         Crust good  
2                            Not tasty texture nasty  
3  Stopped late May bank holiday Rick Steve recom...  
4                    The selection menu great prices

In [ ]:
# Removal of Frequent words

from collections import Counter
cnt = Counter()
for text in df_train["sentence_stopwords"].values:
    for word in text.split():
        cnt[word] += 1
        
cnt.most_common(10)

[('I', 699),
 ('The', 349),
 ('good', 172),
 ('phone', 148),
 ('great', 129),
 ('This', 128),
 ('It', 117),
 ('food', 111),
 ('place', 109),
 ('like', 89)]

In [ ]:
FREQWORDS = set([w for (w, wc) in cnt.most_common(10)])
def remove_freqwords(text):
    """custom function to remove the frequent words"""
    return " ".join([word for word in str(text).split() if word not in FREQWORDS])

# Passing the function to 'sentence_stopwords' and storing into new column 'sentence_freqwords'
df_train["sentence_freqwords"] = df_train["sentence_stopwords"].apply(lambda text: remove_freqwords(text))
df_train.head()

Sentence  Polarity  \
0                           Wow... Loved this place.         1   
1                                 Crust is not good.         0   
2          Not tasty and the texture was just nasty.         0   
3  Stopped by during the late May bank holiday of...         1   
4  The selection on the menu was great and so wer...         1   

                                      sentence_lower  \
0                           wow... loved this place.   
1                                 crust is not good.   
2          not tasty and the texture was just nasty.   
3  stopped by during the late may bank holiday of...   
4  the selection on the menu was great and so wer...   

                                       sentence_punc  \
0                               Wow Loved this place   
1                                  Crust is not good   
2           Not tasty and the texture was just nasty   
3  Stopped by during the late May bank holiday of...   
4  The selection on the menu was great and so wer...   

                                  sentence_stopwords  \
0                                    Wow Loved place   
1                                         Crust good   
2                            Not tasty texture nasty   
3  Stopped late May bank holiday Rick Steve recom...   
4                    The selection menu great prices   

                                  sentence_freqwords  
0                                          Wow Loved  
1                                              Crust  
2                            Not tasty texture nasty  
3  Stopped late May bank holiday Rick Steve recom...  
4                              selection menu prices

In [ ]:
# rare words removal

n_rare_words = 10
RAREWORDS = set([w for (w, wc) in cnt.most_common()[:-n_rare_words-1:-1]])
def remove_rarewords(text):
    """custom function to remove the rare words"""
    return " ".join([word for word in str(text).split() if word not in RAREWORDS])

# Passing the function to 'sentence_freqwords' and storing into new column 'sentence_rarewords'
df_train["sentence_rarewords"] = df_train["sentence_freqwords"].apply(lambda text: remove_rarewords(text))
df_train.head()

Sentence  Polarity  \
0                           Wow... Loved this place.         1   
1                                 Crust is not good.         0   
2          Not tasty and the texture was just nasty.         0   
3  Stopped by during the late May bank holiday of...         1   
4  The selection on the menu was great and so wer...         1   

                                      sentence_lower  \
0                           wow... loved this place.   
1                                 crust is not good.   
2          not tasty and the texture was just nasty.   
3  stopped by during the late may bank holiday of...   
4  the selection on the menu was great and so wer...   

                                       sentence_punc  \
0                               Wow Loved this place   
1                                  Crust is not good   
2           Not tasty and the texture was just nasty   
3  Stopped by during the late May bank holiday of...   
4  The selection on the menu was great and so wer...   

                                  sentence_stopwords  \
0                                    Wow Loved place   
1                                         Crust good   
2                            Not tasty texture nasty   
3  Stopped late May bank holiday Rick Steve recom...   
4                    The selection menu great prices   

                                  sentence_freqwords  \
0                                          Wow Loved   
1                                              Crust   
2                            Not tasty texture nasty   
3  Stopped late May bank holiday Rick Steve recom...   
4                              selection menu prices   

                                  sentence_rarewords  
0                                          Wow Loved  
1                                              Crust  
2                            Not tasty texture nasty  
3  Stopped late May bank holiday Rick Steve recom...  
4                              selection menu prices

In [ ]:
# Stemming

from nltk.stem.porter import PorterStemmer

stemmer = PorterStemmer()
def stem_words(text):
    return " ".join([stemmer.stem(word) for word in text.split()])

# Passing the function to 'sentence_rarewords' and storing into new column 'sentence_stemmed'

df_train["sentence_stemmed"] = df_train["sentence_rarewords"].apply(lambda text: stem_words(text))
df_train.head()



Sentence  Polarity  \
0                           Wow... Loved this place.         1   
1                                 Crust is not good.         0   
2          Not tasty and the texture was just nasty.         0   
3  Stopped by during the late May bank holiday of...         1   
4  The selection on the menu was great and so wer...         1   

                                      sentence_lower  \
0                           wow... loved this place.   
1                                 crust is not good.   
2          not tasty and the texture was just nasty.   
3  stopped by during the late may bank holiday of...   
4  the selection on the menu was great and so wer...   

                                       sentence_punc  \
0                               Wow Loved this place   
1                                  Crust is not good   
2           Not tasty and the texture was just nasty   
3  Stopped by during the late May bank holiday of...   
4  The selection on the menu was great and so wer...   

                                  sentence_stopwords  \
0                                    Wow Loved place   
1                                         Crust good   
2                            Not tasty texture nasty   
3  Stopped late May bank holiday Rick Steve recom...   
4                    The selection menu great prices   

                                  sentence_freqwords  \
0                                          Wow Loved   
1                                              Crust   
2                            Not tasty texture nasty   
3  Stopped late May bank holiday Rick Steve recom...   
4                              selection menu prices   

                                  sentence_rarewords  \
0                                          Wow Loved   
1                                              Crust   
2                            Not tasty texture nasty   
3  Stopped late May bank holiday Rick Steve recom...   
4                              selection menu prices   

                                    sentence_stemmed  
0                                           wow love  
1                                              crust  
2                             not tasti textur nasti  
3  stop late may bank holiday rick steve recommen...  
4                                  select menu price

###  Modeling

In [ ]:
# split data into X and y
X=df_train['sentence_stemmed']
y=df_train['Polarity']

In [ ]:
X

0                                                wow love
1                                                   crust
2                                  not tasti textur nasti
3       stop late may bank holiday rick steve recommen...
4                                       select menu price
                              ...                        
2395               almost song cover girl oldfashion tune
2396    annoy thing cover girl way rita hayworth put p...
2397       unfortun cover girl exampl hollywood use women
2398                            mani everi part quit well
2399                 cinematographi also make look beauti
Name: sentence_stemmed, Length: 2400, dtype: object

In [ ]:
y

0       1
1       0
2       0
3       1
4       1
       ..
2395    0
2396    0
2397    0
2398    1
2399    1
Name: Polarity, Length: 2400, dtype: int64

In [ ]:
# split data into train and test

from sklearn.model_selection import train_test_split 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.30, random_state=123) # 70% train 30% test

print('Training Data :', X_train.shape)
print('Testing Data : ', X_test.shape)

Training Data : (1680,)
Testing Data :  (720,)


In [ ]:
from sklearn.feature_extraction.text import CountVectorizer
cv = CountVectorizer()
X_train_cv = cv.fit_transform(X_train)
X_train_cv.shape

(1680, 2678)

## Logistic Regression

In [ ]:
from sklearn.linear_model import LogisticRegression
lr = LogisticRegression()
lr.fit(X_train_cv, y_train)

LogisticRegression()

In [ ]:
from sklearn import metrics

# transform X_test using CV
X_test_cv = cv.transform(X_test)

# Create a prediction set:
predictions = lr.predict(X_test_cv)

In [ ]:
# classification report
print(metrics.classification_report(y_test,predictions))

              precision    recall  f1-score   support

           0       0.77      0.77      0.77       358
           1       0.78      0.78      0.78       362

    accuracy                           0.78       720
   macro avg       0.77      0.77      0.77       720
weighted avg       0.78      0.78      0.78       720



In [ ]:
# Check AUC
print(metrics.roc_auc_score(y_test,lr.predict_proba(X_test_cv)[:, 1]))

0.8688192536806691


### Random Forest

In [ ]:
from sklearn.ensemble import RandomForestClassifier
rf = RandomForestClassifier()
rf.fit(X_train_cv, y_train)

RandomForestClassifier()

In [ ]:
# predictions set

predictions_rf = rf.predict(cv.transform(X_test))

In [ ]:
# classification report
print(metrics.classification_report(y_test,predictions_rf))

              precision    recall  f1-score   support

           0       0.83      0.69      0.76       358
           1       0.74      0.86      0.80       362

    accuracy                           0.78       720
   macro avg       0.79      0.78      0.78       720
weighted avg       0.79      0.78      0.78       720



In [ ]:
# Check AUC
print(metrics.roc_auc_score(y_test,rf.predict_proba(X_test_cv)[:, 1]))

0.8684951696040001


### Assessing

Use the testing data to measure the accuracy and F1-score of the model.  

### Text Preprocessing on Test Data

In [ ]:
df_test.head()

Sentence  Polarity
0  A good commentary of today's love and undoubte...         1
1  For people who are first timers in film making...         1
2  It was very popular when I was in the cinema, ...         1
3  It's a feel-good film and that's how I felt wh...         1
4  It has northern humour and positive about the ...         1

In [ ]:
text1 = list(df_test['Sentence'])

In [ ]:
# Will apply same pre processing steps as earlier

df_test["sentence_lower"] = df_test["Sentence"].str.lower()
df_test.head()

Sentence  Polarity  \
0  A good commentary of today's love and undoubte...         1   
1  For people who are first timers in film making...         1   
2  It was very popular when I was in the cinema, ...         1   
3  It's a feel-good film and that's how I felt wh...         1   
4  It has northern humour and positive about the ...         1   

                                      sentence_lower  
0  a good commentary of today's love and undoubte...  
1  for people who are first timers in film making...  
2  it was very popular when i was in the cinema, ...  
3  it's a feel-good film and that's how i felt wh...  
4  it has northern humour and positive about the ...

In [ ]:
# Remove punctuations

PUNCT_TO_REMOVE = string.punctuation
def remove_punctuation(text1):
    #custom function to remove the punctuation
    return text1.translate(str.maketrans('', '', PUNCT_TO_REMOVE))

# Passing the function to 'Sentence' and storing into new column 'sentence_punc'

df_test["sentence_punc"] = df_test["Sentence"].apply(lambda text1: remove_punctuation(text1))
df_test.head()

Sentence  Polarity  \
0  A good commentary of today's love and undoubte...         1   
1  For people who are first timers in film making...         1   
2  It was very popular when I was in the cinema, ...         1   
3  It's a feel-good film and that's how I felt wh...         1   
4  It has northern humour and positive about the ...         1   

                                      sentence_lower  \
0  a good commentary of today's love and undoubte...   
1  for people who are first timers in film making...   
2  it was very popular when i was in the cinema, ...   
3  it's a feel-good film and that's how i felt wh...   
4  it has northern humour and positive about the ...   

                                       sentence_punc  
0  A good commentary of todays love and undoubted...  
1  For people who are first timers in film making...  
2  It was very popular when I was in the cinema a...  
3  Its a feelgood film and thats how I felt when ...  
4  It has northern humour and positive about the ...

In [ ]:
# remove stop words

STOPWORDS = set(stopwords.words('english'))
def remove_stopwords(text1):
    """custom function to remove the stopwords"""
    return " ".join([word for word in str(text1).split() if word not in STOPWORDS])

# Passing the function to 'sentence_punc' and storing into new column 'sentence_stopwords'
df_test["sentence_stopwords"] = df_test["sentence_punc"].apply(lambda text1: remove_stopwords(text1))
df_test.head()

Sentence  Polarity  \
0  A good commentary of today's love and undoubte...         1   
1  For people who are first timers in film making...         1   
2  It was very popular when I was in the cinema, ...         1   
3  It's a feel-good film and that's how I felt wh...         1   
4  It has northern humour and positive about the ...         1   

                                      sentence_lower  \
0  a good commentary of today's love and undoubte...   
1  for people who are first timers in film making...   
2  it was very popular when i was in the cinema, ...   
3  it's a feel-good film and that's how i felt wh...   
4  it has northern humour and positive about the ...   

                                       sentence_punc  \
0  A good commentary of todays love and undoubted...   
1  For people who are first timers in film making...   
2  It was very popular when I was in the cinema a...   
3  Its a feelgood film and thats how I felt when ...   
4  It has northern humour and positive about the ...   

                                  sentence_stopwords  
0  A good commentary todays love undoubtedly film...  
1  For people first timers film making I think ex...  
2  It popular I cinema good house good reactions ...  
3       Its feelgood film thats I felt I came cinema  
4   It northern humour positive community represents

In [ ]:
FREQWORDS = set([w for (w, wc) in cnt.most_common(10)])
def remove_freqwords(text1):
    """custom function to remove the frequent words"""
    return " ".join([word for word in str(text1).split() if word not in FREQWORDS])

# Passing the function to 'sentence_stopwords' and storing into new column 'sentence_freqwords'
df_test["sentence_freqwords"] = df_test["sentence_stopwords"].apply(lambda text1: remove_freqwords(text1))
df_test.head()

Sentence  Polarity  \
0  A good commentary of today's love and undoubte...         1   
1  For people who are first timers in film making...         1   
2  It was very popular when I was in the cinema, ...         1   
3  It's a feel-good film and that's how I felt wh...         1   
4  It has northern humour and positive about the ...         1   

                                      sentence_lower  \
0  a good commentary of today's love and undoubte...   
1  for people who are first timers in film making...   
2  it was very popular when i was in the cinema, ...   
3  it's a feel-good film and that's how i felt wh...   
4  it has northern humour and positive about the ...   

                                       sentence_punc  \
0  A good commentary of todays love and undoubted...   
1  For people who are first timers in film making...   
2  It was very popular when I was in the cinema a...   
3  Its a feelgood film and thats how I felt when ...   
4  It has northern humour and positive about the ...   

                                  sentence_stopwords  \
0  A good commentary todays love undoubtedly film...   
1  For people first timers film making I think ex...   
2  It popular I cinema good house good reactions ...   
3       Its feelgood film thats I felt I came cinema   
4   It northern humour positive community represents   

                                  sentence_freqwords  
0  A commentary todays love undoubtedly film wort...  
1  For people first timers film making think exce...  
2       popular cinema house reactions plenty laughs  
3           Its feelgood film thats felt came cinema  
4      northern humour positive community represents

In [ ]:
# rare words removal

n_rare_words = 10
RAREWORDS = set([w for (w, wc) in cnt.most_common()[:-n_rare_words-1:-1]])
def remove_rarewords(text1):
    """custom function to remove the rare words"""
    return " ".join([word for word in str(text1).split() if word not in RAREWORDS])

# Passing the function to 'sentence_freqwords' and storing into new column 'sentence_rarewords'
df_test["sentence_rarewords"] = df_test["sentence_freqwords"].apply(lambda text1: remove_rarewords(text1))
df_test.head()

Sentence  Polarity  \
0  A good commentary of today's love and undoubte...         1   
1  For people who are first timers in film making...         1   
2  It was very popular when I was in the cinema, ...         1   
3  It's a feel-good film and that's how I felt wh...         1   
4  It has northern humour and positive about the ...         1   

                                      sentence_lower  \
0  a good commentary of today's love and undoubte...   
1  for people who are first timers in film making...   
2  it was very popular when i was in the cinema, ...   
3  it's a feel-good film and that's how i felt wh...   
4  it has northern humour and positive about the ...   

                                       sentence_punc  \
0  A good commentary of todays love and undoubted...   
1  For people who are first timers in film making...   
2  It was very popular when I was in the cinema a...   
3  Its a feelgood film and thats how I felt when ...   
4  It has northern humour and positive about the ...   

                                  sentence_stopwords  \
0  A good commentary todays love undoubtedly film...   
1  For people first timers film making I think ex...   
2  It popular I cinema good house good reactions ...   
3       Its feelgood film thats I felt I came cinema   
4   It northern humour positive community represents   

                                  sentence_freqwords  \
0  A commentary todays love undoubtedly film wort...   
1  For people first timers film making think exce...   
2       popular cinema house reactions plenty laughs   
3           Its feelgood film thats felt came cinema   
4      northern humour positive community represents   

                                  sentence_rarewords  
0  A commentary todays love undoubtedly film wort...  
1  For people first timers film making think exce...  
2       popular cinema house reactions plenty laughs  
3           Its feelgood film thats felt came cinema  
4      northern humour positive community represents

In [ ]:
# Stemming

from nltk.stem.porter import PorterStemmer

stemmer = PorterStemmer()
def stem_words(text1):
    return " ".join([stemmer.stem(word) for word in text1.split()])

# Passing the function to 'sentence_rarewords' and storing into new column 'sentence_stemmed'

df_test["sentence_stemmed"] = df_test["sentence_rarewords"].apply(lambda text1: stem_words(text1))
df_test.head()

Sentence  Polarity  \
0  A good commentary of today's love and undoubte...         1   
1  For people who are first timers in film making...         1   
2  It was very popular when I was in the cinema, ...         1   
3  It's a feel-good film and that's how I felt wh...         1   
4  It has northern humour and positive about the ...         1   

                                      sentence_lower  \
0  a good commentary of today's love and undoubte...   
1  for people who are first timers in film making...   
2  it was very popular when i was in the cinema, ...   
3  it's a feel-good film and that's how i felt wh...   
4  it has northern humour and positive about the ...   

                                       sentence_punc  \
0  A good commentary of todays love and undoubted...   
1  For people who are first timers in film making...   
2  It was very popular when I was in the cinema a...   
3  Its a feelgood film and thats how I felt when ...   
4  It has northern humour and positive about the ...   

                                  sentence_stopwords  \
0  A good commentary todays love undoubtedly film...   
1  For people first timers film making I think ex...   
2  It popular I cinema good house good reactions ...   
3       Its feelgood film thats I felt I came cinema   
4   It northern humour positive community represents   

                                  sentence_freqwords  \
0  A commentary todays love undoubtedly film wort...   
1  For people first timers film making think exce...   
2       popular cinema house reactions plenty laughs   
3           Its feelgood film thats felt came cinema   
4      northern humour positive community represents   

                                  sentence_rarewords  \
0  A commentary todays love undoubtedly film wort...   
1  For people first timers film making think exce...   
2       popular cinema house reactions plenty laughs   
3           Its feelgood film thats felt came cinema   
4      northern humour positive community represents   

                                    sentence_stemmed  
0  a commentari today love undoubtedli film worth...  
1    for peopl first timer film make think excel job  
2          popular cinema hous reaction plenti laugh  
3             it feelgood film that felt came cinema  
4                northern humour posit commun repres

In [ ]:
# create a new dataframe

df_test1 = df_test['sentence_stemmed']

In [ ]:
df_test1.head()

0    a commentari today love undoubtedli film worth...
1      for peopl first timer film make think excel job
2            popular cinema hous reaction plenti laugh
3               it feelgood film that felt came cinema
4                  northern humour posit commun repres
Name: sentence_stemmed, dtype: object

In [ ]:
df_test1

0      a commentari today love undoubtedli film worth...
1        for peopl first timer film make think excel job
2              popular cinema hous reaction plenti laugh
3                 it feelgood film that felt came cinema
4                    northern humour posit commun repres
                             ...                        
595                got bore watch jessic lang take cloth
596    unfortun virtu film product work lost regrett ...
597                                    in word embarrass
598                                           except bad
599              all insult one intellig huge wast money
Name: sentence_stemmed, Length: 600, dtype: object

In [ ]:
# transform df_test using CV
df_test_cv = cv.transform(df_test1)

In [ ]:
# Create a prediction set using logistic regression model as it had better auc score

predictions_lr_test = lr.predict(df_test_cv)

In [ ]:
# display predictions

predictions_lr_test


array([1, 1, 1, 0, 0, 0, 1, 0, 1, 1, 1, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1,
       1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 0,
       0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 1, 1, 1, 0, 1, 0,
       0, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0,
       0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1,
       1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 1,
       1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0,
       1, 1, 1, 1, 0, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 0,
       0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 1,
       0, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0, 1, 0, 0, 1, 1, 1,
       0, 1, 0, 0, 0, 1, 1, 0, 1, 1, 0, 0, 0, 0, 0,

In [ ]:
predictions_lr_test_final = pd.DataFrame(predictions_lr_test)

In [ ]:
predictions_lr_test_final

0
0    1
1    1
2    1
3    0
4    0
..  ..
595  0
596  0
597  0
598  0
599  0

[600 rows x 1 columns]

There were two models tested, out of which Logistic Regression gave an AUC score of 0.8688 which was slightly on the higher side when compared to Random Forest.

The Logistic Regression model achieved an accuracy and F1 score of 78%, which is satisfactory but not optimal for making business decisions. Further improvement could be made by increasing the amount of data, performing feature engineering, or exploring alternative machine learning algorithms or ensemble methods. It is important to consider the cost and time required for these tasks.

### The model's predictions might not be completely precise in certain scenarios. The following examples illustrate the reasons behind this.

In [ ]:
# combine predicted score and sentence column

predictions_lr_test_final['Sentence pre']= df_test['sentence_stemmed']

In [ ]:
predictions_lr_test_final['Sentence orginal']= df_test['Sentence']

In [ ]:
predictions_lr_test_final.head()

0                                       Sentence pre  \
0  1  a commentari today love undoubtedli film worth...   
1  1    for peopl first timer film make think excel job   
2  1          popular cinema hous reaction plenti laugh   
3  0             it feelgood film that felt came cinema   
4  0                northern humour posit commun repres   

                                    Sentence orginal  
0  A good commentary of today's love and undoubte...  
1  For people who are first timers in film making...  
2  It was very popular when I was in the cinema, ...  
3  It's a feel-good film and that's how I felt wh...  
4  It has northern humour and positive about the ...

Column 0 are predicted scores

## Two examples that shows model's predictions were incorrect

### 1st example

1 . As per the predictions in the table above, the sentence " It's a feel-good film and that's how I felt when I came out of the cinema! " is labeled 0 but its actually 1.

Although the statement has a very positive tone when taken as a whole, the model may have classified the phrase "came out of the cinema" as negative.

### 2nd Example

In [ ]:
display(predictions_lr_test_final.iloc[5])

0                                        0
Sentence pre                  rather enjoy
Sentence orginal    I rather enjoyed it.  
Name: 5, dtype: object

2. As per the predictions, the sentence "I rather enjoyed it." is labeled 0 but its actually 1.

Several factors could have contributed to the negative labeling instead of positive. Firstly, the sentence is too brief to make a definite assessment, and the statement itself has a negative impact. Moreover, the model might have identified the term "rather" as unfavorable.